sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosBulkItemResponse.java

```diff
@@ -0,0 +1,189 @@
+        this.statusCode = statusCode;
+        this.retryAfter = retryAfter;
+        this.subStatusCode = subStatusCode;
+        this.responseHeaders = responseHeaders;
+        this.cosmosDiagnostics = cosmosDiagnostics;
+    }
+
+    /**
+     * Gets the activity ID that identifies the server request made to execute this operation.
+     *
+     * @return the activity ID that identifies the server request made to execute this operation.
+     */
+    public String getActivityId() {
+        return BatchExecUtils.getActivityId(this.responseHeaders);
+    }
+
+    /**
+     * Gets the entity tag associated with the current item.
+     *
+     * ETags are used for concurrency checking when updating resources.
+     *
+     * @return Entity tag associated with the current item.
+     */
+    public String getETag() {
+        return this.eTag;
+    }
+
+    /**
+     * Gets the request charge as request units (RU) consumed by the current operation.
+     * <p>
+     * For more information about the RU and factors that can impact the effective charges please visit
+     * <a href="https://docs.microsoft.com/en-us/azure/cosmos-db/request-units">Request Units in Azure Cosmos DB</a>
+     *
+     * @return the request charge.
+     */
+    public double getRequestCharge() {
+        return this.requestCharge;
+    }
+
+    /**
+     * Gets the item associated with the current result.
+     *
+     * @param <T> the type parameter
+     *
+     * @param type class type for which deserialization is needed.
+     *
+     * @return item associated with the current result.
+     */
+    public <T> T getItem(final Class<T> type) {
+        T item = null;
+
+        if (this.getResourceObject() != null) {
+            item = new JsonSerializable(this.getResourceObject()).toObject(type);
+        }
+
+        return item;
+    }
+
+    /**
+     * Gets retry after.
+     *
+     * @return the retry after
+     */
+    public Duration getRetryAfterDuration() {
+        return this.retryAfter;
+    }
+
+    /**
+     * Gets sub status code associated with the current result.
+     *
+     * @return the sub status code
+     */
+    public int getSubStatusCode() {
+        return this.subStatusCode;
+    }
+
+    /**
+     * Gets a value indicating whether the current operation completed successfully.
+     *
+     * @return {@code true} if the current operation completed successfully; {@code false} otherwise.
+     */
+    public boolean isSuccessStatusCode() {
+        return 200 <= this.statusCode && this.statusCode <= 299;
+    }
+
+    /**
+     * Gets the HTTP status code associated with the current result.
+     *
+     * @return the status code.
+     */
+    public int getStatusCode() {
+        return this.statusCode;
+    }
+
+    /**
+     * Gets the cosmos diagnostic for this operation.
+     *
+     * @return the CosmosDiagnostics{@link CosmosDiagnostics}
+     */
+    public CosmosDiagnostics getCosmosDiagnostics() {
```

Comment:
Could you paste a sample CosmosDiagnostics for bulk operation somewhere in the PR description? I wonder with all the newly introduced retries how it will look like.

Reply:
Right now in case there is a partial retry, it will have diagnostic for the last successful batch request as we don't have diagnostic operation wise.

If it's permissible to initialize one diagnostic operation wise then we can do some sort of merging there every time a batch request is executed? This will also help in adding diagnostic for getting partition key range id.

Reply:
Let's start with providing the sample here - then we can review whether it is sufficient or how we want to evolve it - makes sense?

Reply:
Sure. So this is the sample for a direct TCP. You can see operation type batch and it's headers here. The response has session token and all.

`{"userAgent":"azsdk-java-cosmos/4.9.0-beta.1 Windows10/10.0 JRE/1.8.0_211","requestLatencyInMs":3533,"requestStartTimeUTC":"2020-11-03T14:09:54.676Z","requestEndTimeUTC":"2020-11-03T14:09:58.209Z","connectionMode":"DIRECT","responseStatisticsList":[{"storeResult":{"storePhysicalAddress":"rntbd://cdb-ms-stage-eastus2-fd4.documents-staging.windows-ppe.net:14013/apps/b5716ed2-296e-4856-89aa-03d65f8dd46b/services/715cd67d-5ed6-4dd0-a256-bc574a605d13/partitions/6fb56c52-6f2f-4a37-a578-199d36bac8d2/replicas/132488385023590763p/","lsn":4,"globalCommittedLsn":3,"partitionKeyRangeId":"1","isValid":true,"statusCode":200,"subStatusCode":0,"isGone":false,"isNotFound":false,"isInvalidPartition":false,"requestCharge":606.67,"itemLSN":-1,"sessionToken":"-1#4","exception":null,"transportRequestTimeline":[{"eventName":"created","startTimeUTC":"2020-11-03T14:09:55.567Z"},{"eventName":"queued","startTimeUTC":"2020-11-03T14:09:55.652Z"},{"eventName":"channelAcquisitionStarted","startTimeUTC":"2020-11-03T14:09:55.652Z"},{"eventName":"pipelined","startTimeUTC":"2020-11-03T14:09:57.185Z"},{"eventName":"transitTime","startTimeUTC":"2020-11-03T14:09:57.256Z"},{"eventName":"received","startTimeUTC":"2020-11-03T14:09:58.195Z"},{"eventName":"completed","startTimeUTC":"2020-11-03T14:09:58.209Z"}],"rntbdRequestLengthInBytes":22674,"rntbdResponseLengthInBytes":43263,"requestPayloadLengthInBytes":22186,"responsePayloadLengthInBytes":42934,"channelTaskQueueSize":0,"pendingRequestsCount":0,"serviceEndpointStatistics":{"availableChannels":0,"acquiredChannels":0,"executorTaskQueueSize":0,"inflightRequests":1,"lastSuccessfulRequestTime":"2020-11-03T14:09:55.600Z","lastRequestTime":"2020-11-03T14:09:55.600Z","createdTime":"2020-11-03T14:09:55.578Z","isClosed":false}},"requestResponseTimeUTC":"2020-11-03T14:09:58.209Z","requestResourceType":"Document","requestOperationType":"Batch"}],"supplementalResponseStatisticsList":[],"addressResolutionStatistics":{"3fa7837d-1dde-11eb-804d-03762526cb03":{"startTimeUTC":"2020-11-03T14:09:55.282Z","endTimeUTC":"2020-11-03T14:09:55.561Z","targetEndpoint":"https://rakkumastage-eastus2.documents-staging.windows-ppe.net:443/addresses/?$resolveFor=dbs%2FwEo8AA%3D%3D%2Fcolls%2FwEo8AJyCb1o%3D%2Fdocs&$filter=protocol%20eq%20rntbd&$partitionKeyRangeIds=1","errorMessage":null,"inflightRequest":false}},"regionsContacted":["https://rakkumastage-eastus2.documents-staging.windows-ppe.net:443/"],"retryContext":{"retryCount":0,"statusAndSubStatusCodes":null,"retryLatency":0},"metadataDiagnosticsContext":{"metadataDiagnosticList":[{"metaDataName":"CONTAINER_LOOK_UP","startTimeUTC":"2020-11-03T14:09:54.676Z","endTimeUTC":"2020-11-03T14:09:55.281Z","durationinMS":605},{"metaDataName":"SERVER_ADDRESS_LOOKUP","startTimeUTC":"2020-11-03T14:09:55.282Z","endTimeUTC":"2020-11-03T14:09:55.561Z","durationinMS":279}]},"serializationDiagnosticsContext":{"serializationDiagnosticsList":[{"serializationType":"ITEM_SERIALIZATION","startTimeUTC":"2020-11-03T14:09:54.676Z","endTimeUTC":"2020-11-03T14:09:54.676Z","durationInMicroSec":0}]},"gatewayStatistics":null,"systemInformation":{"usedMemory":"46861 KB","availableMemory":"1817331 KB","systemCpuLoad":"(2020-11-03T14:09:49.474Z 12.9%), (2020-11-03T14:09:49.474Z 12.9%), (2020-11-03T14:09:49.474Z 12.9%), (2020-11-03T14:09:49.474Z 12.9%), (2020-11-03T14:09:49.588Z 12.9%), (2020-11-03T14:09:54.596Z 8.6%)"},"clientCfgs":{"id":-3,"numberOfClients":4,"connCfg":{"rntbd":"(cto:PT5S, rto:PT5S, icto:PT0S, ieto:PT1H, mcpe:130, mrpc:30, cer:false)","gw":"(cps:1000, rto:PT5S, icto:null, p:false)","other":"(ed: true, cs: false)"},"consistencyCfg":"(consistency: Session, mm: true, prgns: [])"}}`

For a gone exception(changed the pk range in code to get a gone):
Exception:
`PartitionKeyRangeGoneException{userAgent=azsdk-java-cosmos/4.9.0-beta.1 Windows10/10.0 JRE/1.8.0_211, error={"Errors":["The requested resource is no longer available at the server."]}, resourceAddress='null', requestUri='null', statusCode=410, message=["The requested resource is no longer available at the server."], {"userAgent":"azsdk-java-cosmos/4.9.0-beta.1 Windows10/10.0 JRE/1.8.0_211","requestLatencyInMs":0,"requestStartTimeUTC":"2020-11-03T14:48:12.367Z","requestEndTimeUTC":"2020-11-03T14:48:12.367Z","connectionMode":"DIRECT","responseStatisticsList":[],"supplementalResponseStatisticsList":[],"addressResolutionStatistics":{},"regionsContacted":[],"retryContext":{"retryCount":0,"statusAndSubStatusCodes":null,"retryLatency":0},"metadataDiagnosticsContext":{"metadataDiagnosticList":null},"serializationDiagnosticsContext":{"serializationDiagnosticsList":[{"serializationType":"ITEM_SERIALIZATION","startTimeUTC":"2020-11-03T14:48:12.367Z","endTimeUTC":"2020-11-03T14:48:12.367Z","durationInMicroSec":0}]},"gatewayStatistics":null,"systemInformation":{"usedMemory":"59794 KB","availableMemory":"1804398 KB","systemCpuLoad":"(2020-11-03T14:48:01.426Z 20.5%), (2020-11-03T14:48:05.116Z 9.8%), (2020-11-03T14:48:10.120Z 16.0%)"},"clientCfgs":{"id":-1,"numberOfClients":2,"connCfg":{"rntbd":"(cto:PT5S, rto:PT5S, icto:PT0S, ieto:PT1H, mcpe:130, mrpc:30, cer:false)","gw":"(cps:1000, rto:PT5S, icto:null, p:false)","other":"(ed: true, cs: false)"},"consistencyCfg":"(consistency: Session, mm: true, prgns: [])"}}, causeInfo=null, responseHeaders={x-ms-current-replica-set-size=4, x-ms-last-state-change-utc=Fri, 30 Oct 2020 16:31:52.563 GMT, lsn=1, x-ms-request-charge=0.00, x-ms-schemaversion=1.10, x-ms-transport-request-id=2, x-ms-number-of-read-regions=0, x-ms-current-write-quorum=3, x-ms-cosmos-quorum-acked-llsn=1, x-ms-quorum-acked-lsn=1, x-ms-activity-id=98d2de0c-1de3-11eb-8335-5de6f3380958, x-ms-xp-role=1, x-ms-global-Committed-lsn=1, x-ms-item-count=10, x-ms-cosmos-llsn=1, x-ms-serviceversion= version=2.11.0.0, x-ms-substatus=1002}, requestHeaders=null}`

Diagnostic for the same:
`{"userAgent":"azsdk-java-cosmos/4.9.0-beta.1 Windows10/10.0 JRE/1.8.0_211","requestLatencyInMs":0,"requestStartTimeUTC":"2020-11-03T14:48:12.367Z","requestEndTimeUTC":"2020-11-03T14:48:12.367Z","connectionMode":"DIRECT","responseStatisticsList":[],"supplementalResponseStatisticsList":[],"addressResolutionStatistics":{},"regionsContacted":[],"retryContext":{"retryCount":0,"statusAndSubStatusCodes":null,"retryLatency":0},"metadataDiagnosticsContext":{"metadataDiagnosticList":null},"serializationDiagnosticsContext":{"serializationDiagnosticsList":[{"serializationType":"ITEM_SERIALIZATION","startTimeUTC":"2020-11-03T14:48:12.367Z","endTimeUTC":"2020-11-03T14:48:12.367Z","durationInMicroSec":0}]},"gatewayStatistics":null,"systemInformation":{"usedMemory":"59794 KB","availableMemory":"1804398 KB","systemCpuLoad":"(2020-11-03T14:48:01.426Z 20.5%), (2020-11-03T14:48:05.116Z 9.8%), (2020-11-03T14:48:10.120Z 16.0%)"},"clientCfgs":{"id":-1,"numberOfClients":2,"connCfg":{"rntbd":"(cto:PT5S, rto:PT5S, icto:PT0S, ieto:PT1H, mcpe:130, mrpc:30, cer:false)","gw":"(cps:1000, rto:PT5S, icto:null, p:false)","other":"(ed: true, cs: false)"},"consistencyCfg":"(consistency: Session, mm: true, prgns: [])"}}
`

Reply:
The retry logic works in gone, have tested all the test case with a one time gone exception such that if retry attempt is one, keep the valid pk range. But we need to fix the diagnostic such that it tracks this retry and any retry during throttling.